### PR TITLE
MULE-8511 Add support for non blocking processing of flows

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core/src/main/java/org/mule/DefaultMuleEvent.java [hunk 1]
```diff
@@ -118,15 +121,26 @@ public DefaultMuleEvent(MuleMessage message,
             .getDefaultResponseTimeout(), null, null);
     }
 
+    public DefaultMuleEvent(MuleMessage message,
```
Comment (pablokraan) [inline core/src/main/java/org/mule/DefaultMuleEvent.java]: javadoc?

File: core/src/main/java/org/mule/DefaultMuleEvent.java [hunk 2]
```diff
@@ -203,13 +218,25 @@ public DefaultMuleEvent(MuleMessage message,
...
         this.processingTime = ProcessingTime.newInstance(this);
-        this.replyToHandler = null;
+        this.replyToHandler = replyToHandler;
         this.replyToDestination = null;
         this.timeout = timeout;
         this.transacted = false;
         this.synchronous = resolveEventSynchronicity();
     }
 
+    public DefaultMuleEvent(MuleMessage message,
```
Comment (pablokraan) [inline core/src/main/java/org/mule/DefaultMuleEvent.java]: javadoc?

File: core/src/main/java/org/mule/DefaultMuleEvent.java [hunk 3]
```diff
@@ -272,10 +299,20 @@ public DefaultMuleEvent(MuleEvent rewriteEvent, FlowConstruct flowConstruct)
             rewriteEvent.isSynchronous());
     }
 
+    /**
+     * Copy constructor used when ReplyToHandler instance needs switching out
+     * @param rewriteEvent
```
Comment (pablokraan) [inline core/src/main/java/org/mule/DefaultMuleEvent.java]: Add line between description and paramters

File: core/src/main/java/org/mule/DefaultMuleEvent.java [hunk 4]
```diff
@@ -285,9 +322,18 @@ public DefaultMuleEvent(MuleMessage message, MuleEvent rewriteEvent, boolean syn
 
     public DefaultMuleEvent(MuleMessage message, MuleEvent rewriteEvent, boolean synchronus, boolean shareFlowVars)
     {
-        this(message, rewriteEvent, rewriteEvent.getFlowConstruct(), rewriteEvent.getSession(), synchronus, shareFlowVars);
+        this(message, rewriteEvent, rewriteEvent.getFlowConstruct(), rewriteEvent.getSession(), synchronus,
+             shareFlowVars, rewriteEvent.getExchangePattern());
+    }
+
+    public DefaultMuleEvent(MuleMessage message, MuleEvent rewriteEvent, boolean synchronus, boolean shareFlowVars,
```
Comment (pablokraan) [inline core/src/main/java/org/mule/DefaultMuleEvent.java]: javadoc?

File: core/src/main/java/org/mule/api/NonBlockingSupported.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.mule.api.processor.MessageProcessor;
+
+/**
+ * Marker interface for {@link org.mule.api.processor.InterceptingMessageProcessor} and
+ * {@link org.mule.api.processor.MessageRouter} implementations that support non-blocking processing.  Unsupported
+ * implementations will cause Mule to fall back to synchronous blocking processing.  Supporting non blocking means that
+ * i) This components does not require synchronous processing ii) That any response processing (processing done on the
+ * result after invoking a child/next {@link org.mule.api.processor.MessageProcessor will also be processed when running
+ * non-blocking.  One easy way of supporting non-blocking if a components needs to do reponse processing is to extend
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/NonBlockingSupported.java]: reponse > response

File: core/src/main/java/org/mule/api/transport/ErrorHandlingCompletionHandlerReplyToHandlerAdaptor.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.mule.api.CompletionHandler;
+
+/**
+ * Specialized version of {@link CompletionHandlerReplyToHandlerAdaptor} that uses a {@link org.mule.api.exception.MessagingExceptionHandler}
+ * to handle exceptions.
+ */
+public class ErrorHandlingCompletionHandlerReplyToHandlerAdaptor extends CompletionHandlerReplyToHandlerAdaptor
+{
+
+    final private MessagingExceptionHandler messagingExceptionHandler;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/transport/ErrorHandlingCompletionHandlerReplyToHandlerAdaptor.java]: private final

File: core/src/main/java/org/mule/api/transport/ReplyToHandler.java [hunk 7]
```diff
@@ -19,4 +20,14 @@
 {
     void processReplyTo(MuleEvent event, MuleMessage returnMessage, Object replyTo) throws MuleException;
 
+    /**
+     * Processes replyTo in the case an exception errored.  Not all implementations will implement this if for example
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/transport/ReplyToHandler.java]: exception errored > exception occurred?

File: core/src/test/java/org/mule/processor/chain/BlockingProcessorExecutorTestCase.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.processor.chain;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.junit.Assert.*;
```
Comment (pablokraan) [inline core/src/test/java/org/mule/processor/chain/BlockingProcessorExecutorTestCase.java]: Don't use * imports

File: core/src/test/java/org/mule/processor/chain/ProcessorExecutorFactoryTestCase.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.processor.chain;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
```
Comment (pablokraan) [inline core/src/test/java/org/mule/processor/chain/ProcessorExecutorFactoryTestCase.java]: Fix this

File: core/src/test/java/org/mule/api/transport/ErrorHandlingCompletionHandlerReplyToHandlerAdaptorTestCase.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.api.transport;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.junit.Assert.*;
```
Comment (pablokraan) [inline core/src/test/java/org/mule/api/transport/ErrorHandlingCompletionHandlerReplyToHandlerAdaptorTestCase.java]: No * imports

File: core/src/main/java/org/mule/processor/BlockingProcessorExecutor.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        while (continueExecuting())
+        {
+            event = executeNext();
+        }
+        return event;
+    }
+
+    protected boolean continueExecuting()
+    {
+        return getIndex() < processors.size() && event != null && !(event instanceof VoidMuleEvent);
```
Comment (pablokraan) [inline core/src/main/java/org/mule/processor/BlockingProcessorExecutor.java]: Not sure about this: seems to me that its possible to have a race condition in the non-blocking scenario when two different threads got getIndex() < processors.size() on the last processor, but then one of them will use processors.get(getIndex()) with an invalid index (ie: greater than processors.size).

File: core/src/main/java/org/mule/DefaultMuleEvent.java (hunk 1 above)
Comment (dfeist) [inline core/src/main/java/org/mule/DefaultMuleEvent.java]: We have javadoc on some constructors and others now. I'm not sure how valuable javadoc which says "creates MuleEvent" on each constructor is though TBH.  I tend to use javadoc on interfaces always, but in implementations only when method signature is not already self-explanatory.

File: core/src/main/java/org/mule/processor/BlockingProcessorExecutor.java (hunk 11 above)
Comment (dfeist) [inline core/src/main/java/org/mule/processor/BlockingProcessorExecutor.java]: Let's think about this today maybe, I don't think there will be a situation where two different threads are invoking continueExecuting() though.

**Pull request closed without merging**